lib/util.js

```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
+      if (visibleLength === maxLength)
+        break;
+      const index = +key;
+      // Arrays can only have up to 2^32 - 1 entries
+      if (index > 2 ** 32 - 2)
         break;
+      if (i !== index) {
+        if (!numberRegExp.test(key))
+          break;
+        const emptyItems = index - i;
+        const ending = emptyItems > 1 ? 's' : '';
+        const message = `<${emptyItems} empty item${ending}>`;
+        output.push(ctx.stylize(message, 'undefined'));
+        i = index;
+        if (++visibleLength === maxLength)
+          break;
+      }
+      output.push(formatProperty(ctx, value, recurseTimes, key, 1));
+      visibleLength++;
+      i++;
     }
-    output.push(formatProperty(ctx, value, recurseTimes, visibleKeys,
-                               elem, true));
-    visibleLength++;
-    index++;
   }
-  if (index < value.length && visibleLength !== maxLength) {
-    const len = value.length - index;
+  if (i < valLen && visibleLength !== maxLength) {
+    const len = valLen - i;
     const ending = len > 1 ? 's' : '';
     const message = `<${len} empty item${ending}>`;
     output.push(ctx.stylize(message, 'undefined'));
-    index = value.length;
+    i = valLen;
+    if (keyLen === 0)
+      return output;
   }
-  var remaining = value.length - index;
+  const remaining = valLen - i;
   if (remaining > 0) {
     output.push(`... ${remaining} more item${remaining > 1 ? 's' : ''}`);
   }
-  for (var n = 0; n < keys.length; n++) {
-    var key = keys[n];
-    if (typeof key === 'symbol' || !numbersOnlyRE.test(key)) {
-      output.push(formatProperty(ctx, value, recurseTimes, visibleKeys,
-                                 key, true));
+  if (ctx.showHidden && keys[keyLen - 1] === 'length') {
+    // No extra keys
+    output.push(formatProperty(ctx, value, recurseTimes, 'length', 2));
+  } else if (valLen === 0 ||
+    keyLen > valLen && keys[valLen - 1] === `${valLen - 1}`) {
+    // The array is not sparse
+    for (i = valLen; i < keyLen; i++)
+      output.push(formatProperty(ctx, value, recurseTimes, keys[i], 2));
+  } else if (keys[keyLen - 1] !== `${valLen - 1}`) {
+    const extra = [];
+    // Only handle special keys
+    var key;
+    for (i = keys.length - 1; i >= 0; i--) {
+      key = keys[i];
+      if (numberRegExp.test(key) && +key < 2 ** 32 - 1)
+        break;
+      extra.push(formatProperty(ctx, value, recurseTimes, key, 2));
     }
+    for (i = extra.length - 1; i >= 0; i--)
+      output.push(extra[i]);
   }
   return output;
 }
 
+function formatArray(ctx, value, recurseTimes, keys) {
+  const len = Math.min(Math.max(0, ctx.maxArrayLength), value.length);
+  const valLen = value.length;
+  const keyLen = keys.length - (ctx.showHidden ? 1 : 0);
+  if (keyLen !== valLen || keys[keyLen - 1] !== `${valLen - 1}`)
+    return formatSpecialArray(ctx, value, recurseTimes, keys, len, valLen);
 
-function formatTypedArray(ctx, value, recurseTimes, visibleKeys, keys) {
+  const output = [];
```

Comment:
Maybe this could be changed to a "fixed length" Array, since we can determine up-front how many elements there will be (based on `remaining` and `ctx.showHidden`)?

Reply:
I still want to benchmark this at some point but I am actually thinking about doing the opposite in general due to the newest v8 blog post (using a predefined Array size results in the internal holey type and that type will not change even when the array became dense).

Reply:
I thought that was fixed in a previous version of V8 already? /cc @bmeurer

Reply:
Interesting. I benchmarked it and the results are better with `new Array(len)` as expected. The [new blogpost](https://v8project.blogspot.com.br/2017/09/elements-kinds-in-v8.html) says the opposite though...

Anyway - I changed it to `new Array` in almost all places.